### ISSUE TITLE
Transmission | Run as non-root user

### ISSUE THREAD

**Issue by Fourdee** [MichaIng/DietPi#350]
Due to Transmission-daemon service starting before network, and, `dietpi-service`:

```
root@DietPi:~# systemctl status transmission-daemon.service
Process: 470 ExecStart=/etc/init.d/transmission-daemon start (code=exited, status=0/SUCCESS)
root@DietPi:~# systemctl status dietpi-service.service
Process: 483 ExecStart=/etc/init.d/dietpi-service start (code=exited, status=0/SUCCESS)
```

Even though `dietpi-services` successfully disables the transmission-daemon service during install (so dietpi-services can take control), its somehow being re-enabled during reboot.

**Comment by Fourdee** [comment]
notes:

```
root@DietPi:~# find / -type f -name 'transmission-daemon.service'
/var/lib/systemd/deb-systemd-helper-enabled/multi-user.target.wants/transmission-daemon.service
/run/systemd/generator.late/transmission-daemon.service
```

#### Test

Disable Service:

```
root@DietPi:~# systemctl disable transmission-daemon
Synchronizing state for transmission-daemon.service with sysvinit using update-rc.d...
Executing /usr/sbin/update-rc.d transmission-daemon defaults
insserv: warning: current start runlevel(s) (empty) of script `transmission-daemon' overrides LSB defaults (2 3 4 5).
insserv: warning: current stop runlevel(s) (0 1 2 3 4 5 6) of script `transmission-daemon' overrides LSB defaults (0 1 6).
Executing /usr/sbin/update-rc.d transmission-daemon disable
insserv: warning: current start runlevel(s) (empty) of script `transmission-daemon' overrides LSB defaults (2 3 4 5).
insserv: warning: current stop runlevel(s) (0 1 2 3 4 5 6) of script `transmission-daemon' overrides LSB defaults (0 1 6).
```

Reload daemon and reboot:

```
systemctl daemon-reload
reboot
```

Still has an enabled state, and not launched by `dietpi-services`:

```
root@DietPi:~# systemctl status transmission-daemon                             ● transmission-daemon.service - LSB: Start or stop the transmission-daemon.
   Loaded: loaded (/etc/init.d/transmission-daemon)
   Active: active (running) since Sat 2016-05-21 15:06:08 BST; 47s ago
  Process: 466 ExecStart=/etc/init.d/transmission-daemon start (code=exited, status=0/SUCCESS)
```

What the .....?

**Comment by xenfomation** [comment]
Woah.... that's nasty!  Argh...
`
echo "ENABLE_DAEMON=0" > /etc/default/transmission-daemon (or /etc/init.d/transmission-daemon)` 
??

**Comment by Fourdee** [comment]
@xenfomation 
Good idea, `ENABLE_DAEMON=0` prevented the service from starting, but also when running `systemctl start transmission-daemon`

Ok, made my own systemD service for this:

```
rm /etc/init.d/transmission-daemon
rm /etc/systemd/system/transmission-daemon.service
cat << _EOF_ > /etc/systemd/system/transmission-daemon.service
[Unit]
Description=Barebones transmission-daemon service
DefaultDependencies=no

[Service]
Type=oneshot
RemainAfterExit=yes
ExecStart=/usr/bin/transmission-daemon --config-dir /var/lib/transmission-daemon/info
ExecStop=/usr/bin/killall -w transmission-daemon
StandardOutput=tty

[Install]
WantedBy=multi-user.target
_EOF_
systemctl enable transmission-daemon.service
systemctl daemon-reload

dietpi-services disable
```

Problem solved :) Service now starts when called by `dietpi-services start` during boot.

**Comment by Fourdee** [comment]
Completed, will also patch existing systems during update.
Commit: https://github.com/Fourdee/DietPi/commit/09af08c53e9354d9b69e40ebe8470a73ac4b63ee

**Comment by Fourdee** [comment]
Ok, this is still occurring on my C2, but only when using DHCP instead of STATIC IP. So most likely `dietpi-services start` (and transmission) is started before DHCP offer is completed.

Attempted to obtain logfile data. Generates file but doesn't write to it:

```
/usr/bin/transmission-daemon --config-dir /var/lib/transmission-daemon/info --logfile /var/log/transmission.log
```

Not a permissions issue:

```
root@DietPi:~# cat /var/log/transmission.log
root@DietPi:~# ls -lha /var/log/transmission.log
-rwxrwxrwx 1 nobody nogroup 0 May 30 16:30 /var/log/transmission.log
```

**Comment by studioai** [comment]
I solved this problem by editing my transmission-daemon.service like below

```
[Unit]
Description=Barebones transmission-daemon service
DefaultDependencies=no
Requires=network.target
After=dhcpcd.service

[Service]
Type=oneshot
RemainAfterExit=yes
ExecStart=/usr/bin/transmission-daemon --config-dir /var/lib/transmission-daemon/info
ExecStop=/usr/bin/killall -w transmission-daemon
StandardOutput=tty

[Install]
WantedBy=multi-user.target
```

**Comment by Fourdee** [comment]
@studioai 
Nice fix :+1: 

I'll look into making `rc.local` start after dhcpd aswell. Since that service calls `dietpi-services start` (eg transmission) during boot, this should fix transmission:9091 and "missing IP" from banner when in dhcp mode.

Notes:
- Turns out we dont use dhcpd in DietPi. Its dhclient (isc-dhcp-client) , but I cant find a service for it.
- `systemctl enable systemd-networkd-wait-online.service` + network-online.target. No effect.

```
cat << _EOF_ > /etc/systemd/system/rc-local.service
[Unit]
Description=/etc/rc.local Compatibility
Wants=network-online.target
After=dietpi-service.service network-online.target

[Service]
Type=oneshot
ExecStart=/etc/rc.local
StandardOutput=tty
RemainAfterExit=yes

[Install]
WantedBy=multi-user.target
_EOF_
systemctl daemon-reload
```

Lets try dhcpd5 client.
- `apt-get purge isc-dhcp-*`
- `apt-get install dhcpcd5`. No effect.

```
cat << _EOF_ > /etc/systemd/system/rc-local.service
[Unit]
Description=/etc/rc.local Compatibility
Wants=network-online.target
After=dietpi-service.service network-online.target dhcpcd.service

[Service]
Type=oneshot
ExecStart=/etc/rc.local
StandardOutput=tty
RemainAfterExit=yes

[Install]
WantedBy=multi-user.target
_EOF_
systemctl daemon-reload
```

Tempted to write my own while loop in rc.local to check for DHCP completed, before starting `dietpi-services`.

**Comment by Fourdee** [comment]
Fixed: https://github.com/Fourdee/DietPi/commit/b0ca6c7f33f06d6b105ec284d473233d1832a72b
`boot` script which is called by `dietpi-service.service` during boot, will now wait for a valid connection before continuing. Max 10 seconds wait.

This also resolves:
- https://github.com/Fourdee/DietPi/issues/284#issuecomment-217544170

This might resolve:
- https://github.com/Fourdee/DietPi/issues/359

**Comment by Fourdee** [comment]
Did some testing with STATIC IP. IP address not showing in dietpi-banner again. lol...

debugging:
- `while (( ! $(route | awk '{print $4}' | grep -ci -m1 'UG') ));` doesn't get triggered. `route` is providing us with a valid UG result (up, has gateway). Although a `ifconfig` at the same time provides no IP address.
- Removed netplug package. Now while loop gets triggered correctly.

So netplug must be doing something to the adapter/kernel, whilst `route` provides us with "false" information. Fun!

**Comment by Fourdee** [comment]
Ok, workaround:
Netplug is now installed on demand in DietPi-Config when both adapters are enabled (we need netplug to allow for this feature to work correctly). Netplug is uninstalled when only 1 adapter is active.

**Comment by MichaIng** [comment]
@Fourdee 
Read https://dietpi.com/phpbb/viewtopic.php?f=11&t=4858 and found this issue as reason why Transmission is running as root.

Much has changed in between and on current version, Stretch, as well as Jessie, the transmission-daemon.service looks like this, both systemd units now:
```
[Unit]
Description=Transmission BitTorrent Daemon
After=network.target

[Service]
User=debian-transmission
Type=notify
ExecStart=/usr/bin/transmission-daemon -f --log-error
ExecStop=/bin/kill -s STOP $MAINPID
ExecReload=/bin/kill -s HUP $MAINPID

[Install]
WantedBy=multi-user.target
```
- It runs as user `debian-transmission` with safer permissions, default download and config dir/files have ownership accordingly.

I would try to revert using our own service file. There should be no service issues any more, since it is a systemd unit now by default.
Not sure how to handle dietpi_userdata cross access permissions. One idea would be to add "dietpi" user to "debian-transmission" group vice versa and adjust "umask" setting in `/etc/transmission-daemon/settings.json` to e.g. `"umask": 7,` to allow file server access to downloaded files and transmission to write to dietpi_userdata.

**Comment by MichaIng** [comment]
@Fourdee
Solution should be similar to how we run Sonarr/mpd etc now, with own user but dietpi group to allow access to e.g. samba mounts and the other way round have file, created by transmission as dietpi group to allow other members access.

I am thinking, as editing systemd units always breaks updates for those files, maybe we should add a separate grouping task/function to dietpi-software (besides files/folder permissions), and add each of our software run user to dietpi group as well as to each others softwares run user group (at least for those where cross access is required)? Still we need to assure that software creates files always with full group permissions 77X.

**Comment by Fourdee** [comment]
@MichaIng 

Yep, this needs doing, i've set to low priority for now, as the application is functional.

**Comment by MichaIng** [comment]
Related: https://github.com/Fourdee/DietPi/issues/2067#issuecomment-426461043
- Plex has same issue than transmission.

**Comment by MichaIng** [comment]
PR up for changing Plex + Transmission run as group "dietpi": https://github.com/Fourdee/DietPi/pull/2117

**Comment by Fourdee** [comment]
Test:
- 🈯️ Sonarr > Transmission > Plex playback

@MichaIng 

Works a charm, completed 👍

**Comment by joaofl** [comment]
I know you have closed this, but I wanted to ask you what you think about the approach I gave to the problem you mentioned above.
First is, instead of running the daemon with the transmission user, I changed it to dietpi, which makes things much easier later.
And I also made my own copy of the settings file, to which I'm pointing when running the service. This avoids those catchy issues with the config files being copied from one place to another and being overwritten in some cases.
From `/etc/systemd/system/transmission-daemon.service` I have:
`ExecStart=/usr/bin/transmission-daemon --config-dir /mnt/dietpi_userdata/transmission-config`

If you think it makes sense, I can pull this changes.

**Comment by joaofl** [comment]
The complete service config is below:
```
[Unit]
Description=Barebones transmission-daemon service
DefaultDependencies=no

[Service]
Type=oneshot
RemainAfterExit=yes
ExecStart=/usr/bin/transmission-daemon --config-dir /mnt/dietpi_userdata/transmission-config
ExecStop=/usr/bin/killall -w transmission-daemon
StandardOutput=tty
User=dietpi

[Install]
WantedBy=multi-user.target
```

**Comment by Fourdee** [comment]
@joaofl 

Transmission now runs under ```debian-transmission``` user and ```dietpi``` group permissions. Both should allow full access to userdata download directory.

**Comment by MichaIng** [comment]
systemd unit is APT install default now, dietpi group added via separate file: `echo -e '[Service]\nGroup=dietpi' > /etc/systemd/system/transmission-daemon.service.d/dietpi-group.conf`
Config file thus is left in APT install default location: `/etc/transmission-daemon/settings.json`
- As long as we install via APT, I would touch as few things as possible, to allow APT updating everything as required?

One thing is, to have everything in one place, to symlink the config dir to `/mnt/dietpi_userdata/transmission_config`?
